runtime/major_gc.c

```diff
@@ -554,10 +563,136 @@ static void mark_ephe_aux (struct mark_stack *stk, intnat *work,
   }
 }
 
+
+#define Pb_size 256
+#define Pb_min 64
+#define Pb_mask (Pb_size - 1)
+#define Queue_prefetch_distance 4
+
+static uintnat rotate1(uintnat x)
+{
+  return (x << ((sizeof x)*8 - 1)) | (x >> 1);
+}
+
+CAMLnoinline static intnat do_some_marking(intnat work)
+{
+  uintnat pb_enqueued = 0, pb_dequeued = 0;
+  int darkened_anything = 0;
+  value pb[Pb_size];
+  uintnat min_pb = Pb_min;
+  struct mark_stack stk = *Caml_state->mark_stack;
+
+  uintnat young_start = (uintnat)Caml_state->young_start;
+  uintnat half_young_len = ((uintnat)Caml_state->young_end - (uintnat)Caml_state->young_start) >> 1;
+#define Is_block_and_not_young(v) \
+  (((intnat)rotate1((uintnat)v - young_start)) > (intnat)half_young_len)
+#ifdef NO_NAKED_POINTERS
+  #define Is_major_block(v) Is_block_and_not_young(v)
+  //#define Is_major_block(v) (Is_block(v) && !Is_young(v))
+#else
+#define Is_major_block(v) (Is_block_and_not_young(v) && Is_in_heap(v))
+#endif
+
+  while (1) {
+    value *scan, *obj_end, *scan_end;
+
+    if (pb_enqueued > pb_dequeued + min_pb) {
+      /* Dequeue from prefetch buffer */
+      value block = pb[(pb_dequeued++) & Pb_mask];
+      header_t hd = Hd_val(block);
+
+      /* FIXME: Forward_tag */
+      if (Tag_hd(hd) == Forward_tag) {
+
+      } else if (Tag_hd(hd) == Infix_tag) {
+        block -= Infix_offset_val(block);
+        hd = Hd_val(block);
+      }
+
+#ifdef NO_NAKED_POINTERS
+      /* See [caml_darken] for a description of this assertion. */
+      CAMLassert (Is_in_heap (block) || Is_black_hd (hd));
+#endif
+      CAMLassert(Is_white_hd(hd) || Is_black_hd(hd));
+      if (!Is_white_hd (hd)) {
+        /* Already black, nothing to do */
+        continue;
+      }
+      // FIXME work accounting
+      hd = Blackhd_hd (hd);
+      Hd_val (block) = hd;
+      darkened_anything = 1;
+      work--; // header word
+      if (Tag_hd (hd) >= No_scan_tag) {
+        /* Nothing to scan here */
+        work -= Wosize_hd (hd);
+        continue;
+      }
+      scan = Op_val(block);
+      obj_end = scan + Wosize_hd(hd);
+
+      if (Tag_hd (hd) == Closure_tag) {
+        uintnat env_offset = Start_env_closinfo(Closinfo_val(block));
+        work -= env_offset;
+        scan += env_offset;
+      }
+    } else if (work <= 0 || stk.count == 0) {
+      if (min_pb > 0) {
+        // drain pb before quitting
+        min_pb = 0;
+        continue;
+      } else {
+        // done
+        break;
+      }
+    } else {
+      mark_entry m = stk.stack[--stk.count];
+      scan = m.start;
+      obj_end = m.end;
+    }
+
+    scan_end = obj_end;
+    work -= obj_end - scan;
+    if (work < 0) {
+      scan_end += work;
+    }
+
+    for (; scan < scan_end; scan++) {
+      value v = *scan;
+      if (Is_major_block(v)) {
+        if (pb_enqueued == pb_dequeued + Pb_size) {
+          break; /* Prefetch buffer is full */
+        }
+        caml_prefetch(Hp_val(v));
+        caml_prefetch(&Field(v, Queue_prefetch_distance - 1));
```

Comment:
This is fairly subtle and deserves a better explanation in the code. In short: yes, I benchmarked this, yes, it does help. If the object straddles a cacheline, the second prefetch is useful. If it does not, the second prefetch is free because it is coalesced by hardware.